crates/cast/src/lib.rs

```diff
@@ -1448,23 +1448,53 @@ impl SimpleCast {
         Ok(ParseUnits::parse_units(value, unit)?.to_string())
     }
 
-    /// Decodes rlp encoded list with hex data
-    ///
-    /// # Example
+    // Decodes RLP encoded data with validation for canonical integer representation
     ///
+    /// # Examples
     /// ```
     /// use cast::SimpleCast as Cast;
     ///
-    /// assert_eq!(Cast::from_rlp("0xc0").unwrap(), "[]");
-    /// assert_eq!(Cast::from_rlp("0x0f").unwrap(), "\"0x0f\"");
-    /// assert_eq!(Cast::from_rlp("0x33").unwrap(), "\"0x33\"");
-    /// assert_eq!(Cast::from_rlp("0xc161").unwrap(), "[\"0x61\"]");
-    /// assert_eq!(Cast::from_rlp("0xc26162").unwrap(), "[\"0x61\",\"0x62\"]");
+    /// assert_eq!(Cast::from_rlp("0xc0", false).unwrap(), "[]");
+    /// assert_eq!(Cast::from_rlp("0x0f", false).unwrap(), "\"0x0f\"");
+    /// assert_eq!(Cast::from_rlp("0x33", false).unwrap(), "\"0x33\"");
+    /// assert_eq!(Cast::from_rlp("0xc161", false).unwrap(), "[\"0x61\"]");
+    /// assert_eq!(Cast::from_rlp("820002", true).is_err(), true);
+    /// assert_eq!(Cast::from_rlp("820002", false).unwrap(), "\"0x0002\"");
+    /// assert_eq!(Cast::from_rlp("00", true).is_err(), true);
+    /// assert_eq!(Cast::from_rlp("00", false).unwrap(), "\"0x00\"");
     /// # Ok::<_, eyre::Report>(())
     /// ```
-    pub fn from_rlp(value: impl AsRef<str>) -> Result<String> {
+    pub fn from_rlp(value: impl AsRef<str>, as_int: bool) -> Result<String> {
+        fn validate_canonical_int(item: &Item) -> Result<()> {
+            use eyre::bail;
+
+            let bytes = match item {
+                Item::Data(b) => b,
+                _ => return Ok(()),
+            };
+
+            if bytes.is_empty() {
+                return Ok(());
+            }
+
+            // Check for leading zeros
+            if bytes[0] == 0 {
+                bail!("rlp: non-canonical integer (leading zero bytes)");
+            }
+
+            if bytes.len() == 1 && bytes[0] < 0x80 {
+                bail!("rlp: non-canonical size information");
+            }
+
+            Ok(())
+        }
+
         let bytes = hex::decode(value.as_ref()).wrap_err("Could not decode hex")?;
         let item = Item::decode(&mut &bytes[..]).wrap_err("Could not decode rlp")?;
+        if as_int {
+            validate_canonical_int(&item).wrap_err("Non-canonical integer")?;
+        }
+
         Ok(item.to_string())
     }
```

Comment:
I think if `as_int` as true we should parse and return it as `U256`, otherwise as `Item`